src/Chainweb/Pact/Backend/RelationalCheckpointer.hs

```diff
@@ -308,3 +302,42 @@ doGetBlockHistory dbenv blockHeader d = runBlockEnv dbenv $ do
         err -> internalError $
                "readHistoryResult': Expected single row with three columns as the \
                \result, got: " <> T.pack (show err)
+
+getEndTxId :: Database -> BlockHeight -> BlockHash -> IO Int64
+getEndTxId db bhi bha = do
+  r <- qry db
+    "SELECT endingtxid FROM BlockHistory WHERE blockheight = ? and hash = ?;"
+    [SInt $ fromIntegral $ bhi, SBlob $ encode $ bha]
+    [RInt]
+  case r of
+    [[SInt tid]] -> return tid
+    [] -> throwM $ BlockHeaderLookupFailure $ "doGetBlockHistory: not in db: " <>
+          sshow (bhi,bha)
+    _ -> internalError $ "doGetBlockHistory: expected single-row int result, got " <> sshow r
+
+doGetHistoricalLookup
+    :: FromJSON v
+    => Db
+    -> BlockHeader
+    -> Domain k v
+    -> RowKey
+    -> IO (Maybe (TxLog Value))
+doGetHistoricalLookup dbenv blockHeader d k = runBlockEnv dbenv $ do
+  callDb "doGetHistoricalLookup" $ \db -> do
+    endTxId <- getEndTxId db bHeight (_blockHash blockHeader)
```

Comment:
We needed to guard `doGetBlockHistory` because it searches within a range of txIds and `startTxId` tried to get the ending txId of block height -1 (which caused `getEndTxId` to blow up). 
For `doGetHistoricalLookup` we don't have to worry about that because we're searching for balances below an upper bound (i.e. the ending txId of the block we received).